src/meta-srv/src/event_recorder.rs

```diff
@@ -0,0 +1,355 @@
+            .context(SendEventSnafu)
+    }
+
+    // The default event handler implementation sends the received events to the frontend instances.
+    fn build_event_handler(&self) -> Box<dyn EventHandler> {
+        Box::new(DefaultEventHandlerImpl {
+            peer_lookup_service: self.peer_lookup_service.clone(),
+        })
+    }
+}
+
+impl EventRecorderImpl {
+    /// Creates a new event recorder to record important events and persist them to the database.
+    pub fn new(meta_peer_client: MetaPeerClientRef) -> Self {
+        let (tx, rx) = channel(DEFAULT_EVENTS_CHANNEL_SIZE);
+        let peer_lookup_service = Arc::new(MetaPeerLookupService::new(meta_peer_client));
+
+        let mut recorder = Self {
+            tx,
+            peer_lookup_service,
+            _handle: None,
+        };
+
+        let processor = EventProcessor::new(rx, recorder.build_event_handler());
+
+        // Spawn a background task to process the events.
+        let handle = tokio::spawn(async move {
+            processor.process().await;
+        });
+
+        recorder._handle = Some(Arc::new(handle));
+
+        recorder
+    }
+}
+
+// DefaultEventHandlerImpl is the default event handler implementation. It sends the received events to the frontend instances.
+struct DefaultEventHandlerImpl {
+    peer_lookup_service: PeerLookupServiceRef,
+}
+
+#[async_trait]
+impl EventHandler for DefaultEventHandlerImpl {
+    async fn handle(&self, events: Vec<Box<dyn Event>>) -> Result<()> {
+        debug!("Received {} events: {:?}", events.len(), events);
+        // TODO(zyy17): Avoid to create a new database client when handling the events.
+        let database_client = self.build_database_client().await?;
+        let row_inserts = RowInsertRequests {
+            inserts: events.into_iter().map(|e| e.to_row_insert()).collect(),
+        };
+        debug!("Inserting events: {:?}", row_inserts);
+
+        database_client
+            .row_inserts_with_hints(
+                row_inserts,
+                &[
+                    (TTL_KEY, DEFAULT_EVENTS_TABLE_TTL),
+                    (APPEND_MODE_KEY, "true"),
+                ],
+            )
+            .await
+            .context(InsertEventsSnafu)?;
+
+        Ok(())
+    }
+}
+
+impl DefaultEventHandlerImpl {
+    async fn build_database_client(&self) -> Result<Database> {
+        let frontends = self
+            .peer_lookup_service
+            .frontends()
+            .await
+            .context(KvBackendSnafu)?;
+
+        if frontends.is_empty() {
+            return NoAvailableFrontendSnafu.fail();
+        }
+
+        let urls = frontends
+            .into_iter()
+            .map(|peer| peer.addr)
+            .collect::<Vec<_>>();
+
+        debug!("Available frontend addresses: {:?}", urls);
+
+        Ok(Database::new(
+            DEFAULT_CATALOG_NAME,
+            DEFAULT_PRIVATE_SCHEMA_NAME,
+            Client::with_urls(urls),
```

Comment:
It's better to use `with_manager_and_urls` so we can share the `ChannelManager`.